File: app/src/main/java/com/ghstudios/android/features/weapons/detail/WeaponDetailFragment.java
package com.ghstudios.android.features.weapons.detail;

import java.util.List;

import android.arch.lifecycle.ViewModelProviders;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.View;
import android.widget.TextView;

import com.ghstudios.android.AppSettings;
import com.ghstudios.android.MHUtils;
import com.ghstudios.android.components.ColumnLabelTextCell;
import com.ghstudios.android.components.TitleBarCell;
import com.ghstudios.android.data.classes.Weapon;
import com.ghstudios.android.mhgendatabase.R;

/**
 * The superclass of all weapon detail fragments.
 * TODO: Fragments do not lend themselves well to being inherited for unique views.
 * Find an alternative way to handle it like custom view groups, or move more functionality
 * to the superclass
 */
public class WeaponDetailFragment extends Fragment {
    protected static final String ARG_WEAPON_ID = "WEAPON_ID";

    protected WeaponDetailViewModel viewModel;

    private TitleBarCell titleBar;
    private ColumnLabelTextCell rarityCell;
    private ColumnLabelTextCell attackCell;
    private ColumnLabelTextCell element1Cell;
    private ColumnLabelTextCell element2Cell;
    private ColumnLabelTextCell affinityCell;
    private ColumnLabelTextCell slotsCell;

    protected TextView mWeaponDescription, mWeaponDefenseTextView,
            mWeaponDefenseTextTextView,
            mWeaponCreationTextView, mWeaponUpgradeTextView;

    public static WeaponDetailFragment newInstance(long weaponId) {
        Bundle args = new Bundle();
        args.putLong(ARG_WEAPON_ID, weaponId);
        WeaponDetailFragment f = new WeaponDetailFragment();
        f.setArguments(args);
        return f;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = ViewModelProviders.of(this).get(WeaponDetailViewModel.class);

        // Check for a Weapon ID as an argument, and find the weapon
        Bundle args = getArguments();
        if (args != null) {
            long weaponId = args.getLong(ARG_WEAPON_ID, -1);
            viewModel.loadWeapon(weaponId);
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Bind general view elements
        // subclasses implement onCreateView, so we have to do it here instead
        titleBar = view.findViewById(R.id.titlebar);
        rarityCell = view.findViewById(R.id.rare);
        attackCell = view.findViewById(R.id.attack);
        element1Cell = view.findViewById(R.id.element1);
        element2Cell = view.findViewById(R.id.element2);
        affinityCell = view.findViewById(R.id.affinity);
        slotsCell = view.findViewById(R.id.slots);

        viewModel.getWeaponData().observe(this, this::populateWeapon);
        viewModel.getWeaponElementData().observe(this, this::populateElementData);
    }

    protected void populateWeapon(Weapon weapon) {
        titleBar.setIcon(weapon);
        titleBar.setTitleText(weapon.getName());
        titleBar.setAltTitleText(weapon.getJpnName());
        titleBar.setAltTitleEnabled(AppSettings.isJapaneseEnabled());

        rarityCell.setValueText(weapon.getRarityString());
        attackCell.setValueText("" + weapon.getAttack());
        affinityCell.setValueText(weapon.getAffinity()+"%");
        slotsCell.setValueText("" + weapon.getSlotString());

        /*
         * Items below are from old code
         */

        mWeaponDescription.setText(weapon.getDescription());

        if(weapon.getDefense()==0)
        {
            mWeaponDefenseTextTextView.setVisibility(View.GONE);
            mWeaponDefenseTextView.setVisibility(View.GONE);
        }
        else
            mWeaponDefenseTextView.setText("" + weapon.getDefense());


        
        String createCost = "" + weapon.getCreationCost() + "z";
        String upgradeCost = "" + weapon.getUpgradeCost() + "z";
        
        if (createCost.equals("0z")) {
            createCost = "-";
        }
        if (upgradeCost.equals("0z")) {
            upgradeCost = "-";
        }
        
        mWeaponCreationTextView.setText(createCost);
        mWeaponUpgradeTextView.setText(upgradeCost);
    }

    private void populateElementData(List<WeaponElementData> items) {
        element1Cell.setVisibility(View.GONE);
        element2Cell.setVisibility(View.GONE);

        if (items == null) {
            return;
        }

        if (items.size() >= 1) {
            WeaponElementData data = items.get(0);
            element1Cell.setLabelText(data.getElement());
            element1Cell.setValueText(String.valueOf(data.getValue()));
            element1Cell.setVisibility(View.VISIBLE);
        }

        if (items.size() >= 2) {
            WeaponElementData data = items.get(1);
            element2Cell.setLabelText(data.getElement());
            element2Cell.setValueText(String.valueOf(data.getValue()));
            element2Cell.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/ghstudios/android/features/weapons/detail/WeaponDetailFragment.kt
package com.ghstudios.android.features.weapons.detail

import androidx.lifecycle.Observer
import android.os.Bundle
import androidx.constraintlayout.widget.Group
import androidx.fragment.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider

import com.ghstudios.android.ClickListeners.ItemClickListener
import com.ghstudios.android.components.ItemRecipeCell
import com.ghstudios.android.data.classes.Component
import com.ghstudios.android.data.classes.Weapon
import com.ghstudios.android.mhgendatabase.R
import com.ghstudios.android.mhgendatabase.databinding.FragmentWeaponDetailBinding
import com.ghstudios.android.util.applyArguments


/**
 * Fragment that displays data information for a weapon.
 * Defers the internal data inflation and population to a subclass of WeaponDetailViewHolder.
 */
class WeaponDetailFragment : Fragment() {
    private val TAG = javaClass.name

    companion object {
        private const val ARG_WEAPON_ID = "WEAPON_ID"

        @JvmStatic fun newInstance(weaponId: Long): WeaponDetailFragment {
            return WeaponDetailFragment().applyArguments {
                putLong(ARG_WEAPON_ID, weaponId)
            }
        }
    }

    private lateinit var binding: FragmentWeaponDetailBinding

    /**
     * Returns the viewmodel owned by the activity, which has already loaded weapon data
     */
    private val viewModel by lazy {
        ViewModelProvider(activity!!).get(WeaponDetailViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentWeaponDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.weaponData.observe(viewLifecycleOwner, Observer(::populateWeapon))
        viewModel.createComponentData.observe(viewLifecycleOwner, Observer {
            populateCreateComponents(view, it)
        })
        viewModel.improveComponentData.observe(viewLifecycleOwner, Observer {
            populateUpgradeComponents(view, it)
        })
    }

    /**
     * Internal helper to populate the view with weapon data, and inflate the appropriate sub-view
     * based on the weapon type.
     */
    private fun populateWeapon(weapon: Weapon?) {
        if (weapon == null) return

        with(binding.titlebar) {
            setIcon(weapon)
            setTitleText(weapon.name)
            setAltTitleText(getString(R.string.value_rare, weapon.rarityString))
        }
        with(binding) {
            weaponDescription.text = weapon.description
            weaponCostCreate.text = "" + weapon.creationCost + "z"
            weaponCostUpgrade.text = "" + weapon.upgradeCost + "z"
        }

        // inflate the subview, depending on weapon type
        val weaponDataContainer = view!!.findViewById<ViewGroup>(R.id.weapon_detail_view)
        weaponDataContainer.removeAllViews()
        val weaponDataView = when (weapon.wtype) {
            Weapon.BOW -> WeaponBowDetailViewHolder(weaponDataContainer)
            Weapon.LIGHT_BOWGUN, Weapon.HEAVY_BOWGUN -> WeaponBowgunDetailViewHolder(weaponDataContainer)
            else -> WeaponBladeDetailViewHolder(weaponDataContainer)
        }
        weaponDataView.bindWeapon(weapon)
    }

    /**
     * Internal helper to populate the craft section with the correct components.
     * Use as a callback.
     */
    private fun populateCreateComponents(view: View, components: List<Component>?) {
        val section = view.findViewById<Group>(R.id.create_section)
        val recipeView = view.findViewById<ItemRecipeCell>(R.id.create_recipe)
        populateRecipe(view, section, recipeView, components)
    }

    /**
     * Internal helper to populate the upgrade section with the correct components.
     * Use as a callback.
     */
    private fun populateUpgradeComponents(view: View, components: List<Component>?) {
        val section = view.findViewById<Group>(R.id.upgrade_section)
        val recipeView = view.findViewById<ItemRecipeCell>(R.id.upgrade_recipe)
        populateRecipe(view, section, recipeView, components)
    }

    /**
     * Internal function to populate an arbitrary recipe
     * @param section Section to show or hide based on whether components is empty or not
     * @param recipeView The object to populate with recipe components
     * @param components The actual recipe items
     */
    private fun populateRecipe(view: View, section: Group, recipeView: ItemRecipeCell, components: List<Component>?) {
        if (components == null || components.isEmpty()) {
            Log.i(TAG, "Recipe has zero components")
            section.visibility = View.GONE
            return
        }

        Log.i(TAG, "Populating recipe with components")
        section.visibility = View.VISIBLE
        for (component in components) {
            val item = component.component
            val itemCell = recipeView.addItem(item, item?.name, component.quantity, false)
            itemCell.setOnClickListener(ItemClickListener(context!!, item!!))
        }

        // Workaround for an intermittant bug with ConstraintLayout's group (where it will sometimes continue being invisible)
        // if we often need to use this workaround, it may be better to either not use group or make an extension for this
        for (id in section.referencedIds) {
            view.findViewById<View>(id)?.visibility = View.VISIBLE
        }
    }
}
